research/2019-12-16_searching_with_the_calm_api.md

```diff
@@ -0,0 +1,241 @@
+          #       </soap:Body>
+          #     </soap:Envelope>
+          #
+          result_root = ET.fromstring(summary_header_resp.content)
+
+          result_string = result_root.find(
+              "./"
+              "{http://schemas.xmlsoap.org/soap/envelope/}Body/"
+              "{http://ds.co.uk/cs/webservices/}SummaryHeaderResponse/"
+              "{http://ds.co.uk/cs/webservices/}SummaryHeaderResult"
+          ).text
+
+          # The result is of the form:
+          #
+          #     <?xml version="1.0" encoding="ISO-8859-1"?><SummaryList><Summary>
+          #       <RecordType>Component</RecordType>
+          #       <IDENTITY></IDENTITY>
+          #       ...
+          #     </Summary></SummaryList>
+          #
+          summary_root = ET.fromstring(result_string).find(".//Summary")
+          yield summary_root
+      ```
+    </details>
+
+The Search request is attached to a persistent "session" -- somehow the API knows you want the results from a given Search request (cookies, maybe).
+I was using a persistent `requests.Session()` in Python and that seemed to work fine.
+
+Two possible approaches:
+
+*   **I want to get everything.**
+
+    Make a Search request with database `Catalog` and search term `*`.
+    That tells you how many results there are, then iterate from `0 to (hit count – 1)` (the results are 0-indexed), and call SummaryHeader for each.
+
+*   **I want to get recently updated stuff.**
+
+    You don't want to make a SummaryHeader request for each record -- that's slow and unnecessary.
+
+    Make a Search request with database `Catalog` and search term `*`.
+    Iterate over that number in batches with Overview, inspecting the Modified field.
+    For records that have been modified recently, call SummaryHeader to get the full record.
+
+
+
+## Candy bag of other thoughts
+
+*   The CALM API is *only* available over HTTP, not HTTPS.
+    We should fix that if we want to use it over the public Internet.
+
+*   The CALM API returns results as XML, but the script below serialises them as JSON.
+    This is fine for experiments, but be careful if using this data in anger.
+    For example, I've seen CALM return XML of the form:
+
+    ```xml
+    <?xml version="1.0" encoding="ISO-8859-1"?><SummaryList><Summary>
+      <AltRefNo>LE/MON/1</AltRefNo>
+      ...
+    </xml>
+    ```
+
+    and
+
+    ```xml
+    <?xml version="1.0" encoding="ISO-8859-1"?><SummaryList><Summary>
+      <AltRefNo>LE/MON/1</AltRefNo>
+      <AltRefNo>LE/MON/1</AltRefNo>
+      ...
+    </xml>
+    ```
+
+    One would serialise to JSON as a string, the other a list.
+
+    This might indicate an error somewhere in CALM that we want to fix (and now we have the API, we can find all those places!)
+
+*   AFAICT, the CALM API is read-only -- so there's less risk of a coding error triggering irreparable damage to the CALM database.
+    (I thought I'd seen that written down somewhere, but I can't find the reference now.)
```

Comment:
Saying this - I wonder if there's some long running queries we could run to lock it up?

Reply:
Probably, but I don’t know what those are. 😉